mdata/store_cassandra.go

```diff
@@ -77,9 +78,37 @@ type cassandraStore struct {
 	session     *gocql.Session
 	writeQueues []chan *ChunkWriteRequest
 	readQueue   chan *ChunkReadRequest
+	ttlTables   map[uint32]string
 }
 
-func NewCassandraStore(addrs, keyspace, consistency, CaPath, Username, Password, hostSelectionPolicy string, timeout, readers, writers, readqsize, writeqsize, retries, protoVer int, ssl, auth, hostVerification bool) (*cassandraStore, error) {
+func (c *cassandraStore) ensureTtlTables(ttls []uint32, keyspace string) error {
+	for _, ttl := range ttls {
+		// calculate table name and window
+		var window int = 1
+		for tmp := ttl / (24 * 60 * 60); tmp > 1; tmp = tmp / 2 {
+			window *= 2
+		}
+		table_name := fmt.Sprintf("metric_%d", window)
+
+		// table isn't know to exist, ensure it exists
+		err := c.session.Query(fmt.Sprintf(table_schema, keyspace, table_name, window)).Exec()
```

Comment:
The window size and the table suffix need to be different.

The general rule of thumb with Cassandra is to try and limit tables to 10s of SSTABLES. I think setting the window appropriately to get no more then 20sstables is a good fit for us.

window_size := (window/20) +1
Once the data is compacted into our "window_size" sstable, it is never compacted again (this means tombstoned data is not deleted). So by having up to 20 sstables, we only incur a 5-10% overhead before data past its ttl is deleted, which is nice.